File: frontend/src/controller/wails/boundary-golden-loader.ts
/**
 * boundary-golden-loader.ts
 *
 * frontend 側の共有 golden 読み込み helper。
 * vitest 実行時に Node.js fs を使い、backend 側と同一の物理ファイル
 * (internal/apitest/testdata/boundary/master_dictionary/*.golden.json) を読む。
 *
 * 両側が同一ファイルを参照することで、片側のみ golden を書き換えた場合に
 * もう片側のテストが失敗する検出経路が成立する。
 */

import { readFileSync } from "node:fs"
import { fileURLToPath, URL } from "node:url"

/**
 * BoundaryGoldenInput は golden ファイルの入力相当の説明を表す。
 */
export interface BoundaryGoldenInput {
  description: string
  method: string
  case: string
  request: unknown
}

/**
 * BoundaryGoldenFile は golden JSON ファイルのトップレベル構造。
 * input は入力相当の説明、expected は期待される応答 DTO field 値。
 */
export interface BoundaryGoldenFile<TExpected = unknown> {
  input: BoundaryGoldenInput
  expected: TExpected
}

/**
 * goldenBasePath は frontend/src/controller/wails/ から
 * internal/apitest/testdata/boundary/master_dictionary/ への相対パスを
 * import.meta.url を使って絶対パスに解決した基底ディレクトリ。
 */
const goldenBasePath = fileURLToPath(
  new URL(
    "../../../../internal/apitest/testdata/boundary/master_dictionary",
    import.meta.url
  )
)

/**
 * loadBoundaryGolden は指定した名前の golden ファイルを読み込み、
 * TExpected 型にキャストした BoundaryGoldenFile を返す。
 *
 * @param name - ファイル名のみ（例: "list_normal.golden.json"）
 */
export function loadBoundaryGolden<TExpected = unknown>(
  name: string
): BoundaryGoldenFile<TExpected> {
  const filePath = `${goldenBasePath}/${name}`
  const raw = readFileSync(filePath, "utf-8")
  return JSON.parse(raw) as BoundaryGoldenFile<TExpected>
}

File: internal/apitest/boundary_golden_loader.go
// Package apitest provides shared test infrastructure for Wails controller boundary tests.
// This file provides the boundary golden loader used by both backend boundary tests
// and as the canonical path reference for frontend boundary tests.
package apitest

import (
	"embed"
	"encoding/json"
	"fmt"
)

// goldenFS embeds all golden JSON files under testdata/boundary/master_dictionary/.
//
//go:embed testdata/boundary/master_dictionary/*.golden.json
var goldenFS embed.FS

// BoundaryGoldenInput describes the input side of a golden case.
type BoundaryGoldenInput struct {
	Description string          `json:"description"`
	Method      string          `json:"method"`
	Case        string          `json:"case"`
	Request     json.RawMessage `json:"request"`
}

// BoundaryGoldenFile is the top-level structure of a single golden JSON file.
// Input holds the input-side description; Expected holds the expected DTO field values
// as a raw JSON value so each test can decode it into the appropriate response DTO type.
type BoundaryGoldenFile struct {
	Input    BoundaryGoldenInput `json:"input"`
	Expected json.RawMessage     `json:"expected"`
}

// LoadBoundaryGolden reads the named golden file from the embedded FS and returns the parsed
// BoundaryGoldenFile. The name must be the bare file name (e.g. "list_normal.golden.json").
// The Expected field is left as raw JSON; callers decode it into the appropriate type.
func LoadBoundaryGolden(name string) (BoundaryGoldenFile, error) {
	path := "testdata/boundary/master_dictionary/" + name
	data, err := goldenFS.ReadFile(path)
	if err != nil {
		return BoundaryGoldenFile{}, fmt.Errorf("load boundary golden %q: %w", name, err)
	}

	var gf BoundaryGoldenFile
	if err := json.Unmarshal(data, &gf); err != nil {
		return BoundaryGoldenFile{}, fmt.Errorf("parse boundary golden %q: %w", name, err)
	}
	return gf, nil
}

// MustLoadBoundaryGolden is like LoadBoundaryGolden but panics on error.
// Use only in test helpers where a missing golden file is always a programming error.
func MustLoadBoundaryGolden(name string) BoundaryGoldenFile {
	gf, err := LoadBoundaryGolden(name)
	if err != nil {
		panic(err)
	}
	return gf
}
